nav2_dwb_controller/dwb_critics/test/obstacle_footprint_test.cpp

```diff
@@ -0,0 +1,255 @@
+  }
+
+  double lineCost(int x0, int x1, int y0, int y1)
+  {
+    return dwb_critics::ObstacleFootprintCritic::lineCost(x0, x1, y0, y1);
+  }
+};
+
+geometry_msgs::msg::Point rotate_origin(geometry_msgs::msg::Point p, double angle)
+{
+  double s = sin(angle);
+  double c = cos(angle);
+
+  // rotate point
+  double xnew = p.x * c - p.y * s;
+  double ynew = p.x * s + p.y * c;
+
+  p.x = xnew;
+  p.y = ynew;
+
+  return p;
+}
+
+geometry_msgs::msg::Point p(double x, double y)
+{
+  geometry_msgs::msg::Point p;
+  p.x = x;
+  p.y = y;
+  return p;
+}
+
+// Variables
+double footprint_size_x_half = 1.8;
+double footprint_size_y_half = 1.6;
+
+std::vector<geometry_msgs::msg::Point> getFootprint()
+{
+  std::vector<geometry_msgs::msg::Point> footprint;
+  footprint.push_back(p(footprint_size_x_half, footprint_size_y_half));
+  footprint.push_back(p(footprint_size_x_half, -footprint_size_y_half));
+  footprint.push_back(p(-footprint_size_x_half, -footprint_size_y_half));
+  footprint.push_back(p(-footprint_size_x_half, footprint_size_y_half));
+  return footprint;
+}
+
+TEST(ObstacleFootprint, GetOrientedFootprint)
+{
+  double theta = 0.1234;
+
+  std::vector<geometry_msgs::msg::Point> footprint_before = getFootprint();
+
+  std::vector<geometry_msgs::msg::Point> footprint_after;
+  geometry_msgs::msg::Pose2D pose;
+  pose.theta = theta;
+  footprint_after = dwb_critics::getOrientedFootprint(pose, footprint_before);
+
+  uint i;
+  for (i = 0; i < footprint_before.size(); i++) {
+    ASSERT_EQ(rotate_origin(footprint_before[i], theta), footprint_after[i]);
+  }
+
+  theta = 5.123;
+  pose.theta = theta;
+  footprint_after = dwb_critics::getOrientedFootprint(pose, footprint_before);
+
+  for (unsigned int i = 0; i < footprint_before.size(); i++) {
+    ASSERT_EQ(rotate_origin(footprint_before[i], theta), footprint_after[i]);
+  }
+}
+
+TEST(ObstacleFootprint, Prepare)
+{
+  std::shared_ptr<dwb_critics::ObstacleFootprintCritic> critic =
+    std::make_shared<dwb_critics::ObstacleFootprintCritic>();
+
+  auto node = nav2_util::LifecycleNode::make_shared("costmap_tester");
+
+  auto costmap_ros = std::make_shared<nav2_costmap_2d::Costmap2DROS>("test_global_costmap");
+  // costmap_ros->activate();
+  costmap_ros->configure();
+
+  critic->initialize(node, "name", "ns", costmap_ros);
+
+  geometry_msgs::msg::Pose2D pose;
+  nav_2d_msgs::msg::Twist2D vel;
+  geometry_msgs::msg::Pose2D goal;
+  nav_2d_msgs::msg::Path2D global_plan;
+
+  // no footprint set in the costmap. Prepare should return false;
+  std::vector<geometry_msgs::msg::Point> footprint;
+  costmap_ros->setRobotFootprint(footprint);
+  ASSERT_FALSE(critic->prepare(pose, vel, goal, global_plan));
+
+  costmap_ros->setRobotFootprint(getFootprint());
+  ASSERT_TRUE(critic->prepare(pose, vel, goal, global_plan));
+
+  double epsilon = 0.01;
+  // If the robot footprint goes of the map, it should trow an exception
+  pose.x = footprint_size_x_half;  // This gives an error
```

Comment:
You have these 4 cases here, I suspect they're each covering a different throw in the code, can you give a comment on them as to which are trying to throw from which reasons?

Reply:
They are not really different throws, but they put the footprint of the map in different directions.